src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -358,14 +342,24 @@ public static DecimalFloatingPointString FromSource(string source)
                     int firstExponent = i;
                     int lastExponent = i;
                     while (i < source.Length && source[i] >= '0' && source[i] <= '9') lastExponent = ++i;
-                    int exponentMagnitude = int.Parse(source.Substring(firstExponent, lastExponent - firstExponent));
-                    if (exponentSign == '-')
+
+                    int exponentMagnitude = 0;
+
+                    if (int.TryParse(source.Substring(firstExponent, lastExponent - firstExponent), out exponentMagnitude) &&
+                        exponentMagnitude <= MAX_EXP_QUAD)
                     {
-                        exponent -= exponentMagnitude;
+                        if (exponentSign == '-')
+                        {
+                            exponent -= exponentMagnitude;
+                        }
+                        else
+                        {
+                            exponent += exponentMagnitude;
+                        }
                     }
                     else
                     {
-                        exponent += exponentMagnitude;
+                        exponent = exponentSign == '-' ? -MAX_EXP_QUAD : MAX_EXP_QUAD;
```

Comment:
Ignore this comment.  I hadn't finished my coffee and was reading this as `a = b = ...` instead of `a = b == ... ? ...`.